[FIX] Recommend SI units formatting to adhere to CMIXF-12

sappelhoff opened the pull request:
replaces #151 as proposed by @satra , please see discussion therein!

Here is a summary:

- In BIDS we accept SI units only (...and in very special circumstances, custom units - but they'd need to be documented in JSON files)
- Currently, we have nowhere in the spec clarified how the FORMATTING of the SI units should be done (although some users and developers have **assumed** that the [units appendix](https://bids-specification.readthedocs.io/en/stable/99-appendices/05-units.html) shows the required formatting)
- Thus in practice, users have applied either ASCII formatting (e.g., "u" instead of  "µ") or unicode formatting (µ) ... with the problem being that some unicodes **look** the same, while being different (e.g., µ and μ ... micro vs. mu)

This PR is intended to explicitly state how we want the FORMATTING of SI units to be. Namely, we prefer ASCII formatting according to MIXF (metric interchange format), see --> http://people.csail.mit.edu/jaffer/MIXF/CMIXF-12.

For backward compatibility, we will allow unicode characters, but further clarify the exact unicode we expect (e.g, U+00B5 for the micro sign µ).

Please see the proposed changes in this PR and approve or discuss!


more recent summary: https://github.com/bids-standard/bids-specification/pull/411#issuecomment-617036529

# To do

- [x] add paragraphs to spec that teach users how we expect the formatting to be 
- [x] discuss with community and reach consensus on whether CMIXF is a good idea
- [x] get some kind of a mixf validator (we got https://github.com/sensein/cmixf by Satra)
- [ ] stop bids-validator from units validation how it is currently implemented (https://github.com/bids-standard/bids-validator/issues/969)
- [ ] make bids-validator check MIXF units (https://github.com/bids-standard/bids-validator/issues/969) ... and warn if the units are not compliant (but no error, we only RECOMMEND)

effigies commented on src/99-appendices/05-units.md:
```diff
@@ -1,39 +1,44 @@
+may be found. The primary reason for choosing an ASCII representation, is that
+certain characters can have multiple unicode representations that look very
+similar. The BIDS specification will allow only the following unicode symbols for
+backwards compatibility: U+00B5 (µ) and U+00B0 (°) and U+2126 (Ω).
```
I might rewrite this addition:

```
The [CMIXF-12](https://people.csail.mit.edu/jaffer/MIXF/CMIXF-12) convention for encoding units
is RECOMMENDED to achieve maximum portability and limited variability of representation.
Earlier versions of the BIDS standard listed the following Unicode symbols, and these are still
included for backwards compatibility: `U+00B5` (µ) and `U+00B0` (°) and `U+2126` (Ω).
It is REQUIRED that units be CMIXF-12 compliant or among these three Unicode characters.
```

sappelhoff commented:
Thanks @effigies, I think your wording was more unambiguous and I added it.

I think we could still clarify a bit more with the units, before we try to get this PR in.

**1. For encoding a unit, can a user pick any combination of:**

- prefix and unit,  e.g. `microV`, `microvolt`
- symbol and unit, e.g. `uV`, `µV`, `uvolt`, `µvolt`
- factor and unit, e.g. `1e-6V`, `1e-6volt` (probably not)

... or do we require e.g. SYMBOL and UNIT

**2. (implied in 1.) Can a user write out the complete unit? E.g., `volt` OR `V` are both permissible?**

**3. Regarding lower and upper case, I assume that we go with how it's handled in CMIXF-12, yet it may be good to make this more obvious with an additional sentence.**

sappelhoff commented:
let me just make a proposal that's easier to criticize than my 3 questions above.

Let's add this sentence right before the table:

```
Units MUST consist of the `unit symbol` with an optionally accompanying
`prefix symbol` (see table below). Appropriate upper- or lower- casing MUST
be applied as declared by CMIXF-12.

Examples:
1. `uV` or `µV` are permissible, but NOT: `microV`, `µvolt`, `1e-6V`, etc.
1. 
```

feel free to add an example. Or adjust this text to be more clear.

satra commented:
should we clarify that combinations of units are allowed if required. For example slew rate (`T/m/s`).

effigies commented:
@sappelhoff I agree in principle with what you are proposing, but we should also note that CMIXF-12 does have conventions for derived units, including parentheses, exponents, multiplication and division.

sappelhoff commented:
reminder: this PR will also need to be coordinated with #391

sappelhoff commented:
This is a biggish change, I'd like to get some more feedback on this @yarikoptic @robertoostenveld @nicholst @oesteban

sappelhoff commented:
I think this issue deserves more attention, so let me write a summary for everbody to catch up and offer their opinion.

Selective tag list (please tag more people that I forgot, I also pinned this issue to make it more visible):

@yarikoptic @robertoostenveld @nicholst @oesteban @franklin-feingold @emdupre @Remi-Gau @satra @VisLab @dorahermes @effigies 


**This issue is about how to *format* units in BIDS.**

# Brief history

- Using the [CMIXF-12 formatting for units](https://people.csail.mit.edu/jaffer/MIXF/CMIXF-12) was first proposed by @satra  14 months ago, but we did not act any further on it (https://github.com/bids-standard/bids-specification/pull/151)
- The issue more recently (~3 months ago) came up again, because the validator (cc @rwblair ) started validating units (see: https://github.com/bids-standard/bids-validator/pull/885)

# The problem 

I think it's great that the validator now checks units ... but this also put the finger on the wound in our spec: Namely, that the way units are currently defined in BIDS is _unclear_ with regards to _formatting_.

We clearly state that [we RECOMMEND SI units](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#units) but we are *unclear* as to how we want these units to be formatted.

Some users have taken a look at our [units appendix](https://bids-specification.readthedocs.io/en/stable/99-appendices/05-units.html) and deduced that they can make use of any combination of:

- unit name
- unit symbol
- prefix name
- prefix symbol

For example: µV, µVolt, microVolt, microV

and this is, what the validator currently checks](https://github.com/bids-standard/bids-validator/blob/master/bids-validator/utils/unit.js).

Yet nowhere in the spec to we REQUIRE this behavior. So other users have (arguably reasonably) formatted their SI units in some other ways. E.g., uV to reflect "micro volts".

# Current situation and where to go from here

This brings us to the current situation: The specification is unclear with how to format units, but the validator is strict in what it accepts as a valid unit.

This PR is intended to clarify how we want to have units formatted in BIDS. If you care about this problem, please offer your opinion ... and if you know somebody who might care, please refer them to this place :-) :rocket:

nicholst commented:
I like the clarity of MIXF but I can understand that doesn't totally specify things.

This seems like an area where others have put some thought, though... I don't have time to digest, but it seems the following will be helpful...

* "[Recommended Unit Symbols, SI Prefixes, and Abbreviations](https://www.ewh.ieee.org/soc/ias/pub-dept/abbreviation.pdf)", IEEE document
* "[SCC 14 Conventions for Metrication of IEEE Standards](https://standards.ieee.org/content/dam/ieee-standards/standards/web/governance/revcom/scc14.pdf)", IEEE Standards Coordinating Committee document
* IEEE. (2002). IEEE/ASTM Standard for Use of the International System of Units (SI): The Modern Metric System. IEEE/ASTM SI 10-2002 (Revision of IEEE/ASTM SI 10-1997). https://doi.org/10.1109/IEEESTD.2002.94225

The last one, an IEEE/ANSI standard, even has a section titled "Application of SI prefixes".

I'm hoping we can simply direct readers to one of these standards to say "do it that way, but render it as per MIXF".

alexrockhill commented:
Here is a minimally reproducible example (@jasmainak @sappelhoff) which gives the error screenshot below.

<img width="359" alt="Screen Shot 2020-04-22 at 12 23 00 PM" src="https://user-images.githubusercontent.com/13473576/80047811-463b4500-84c3-11ea-9ab6-e31bd7efc367.png">

```
import os.path as op
import mne

from mne_bids import make_bids_basename, write_raw_bids

base_path = op.join(op.dirname(mne.__file__), 'io')
subject_id = '01'
subject_id2 = '02'
session_id = '01'
run = '01'
acq = '01'
run2 = '02'
task = 'testing'

bids_basename = make_bids_basename(
    subject=subject_id, session=session_id, run=run, acquisition=acq,
    task=task)

bids_root = './bids_dataset'
data_path = op.join(mne.datasets.testing.data_path(), 'EDF')
raw_fname = op.join(data_path, 'test_reduced.edf')

raw = mne.io.read_raw_edf(raw_fname)

write_raw_bids(raw, bids_basename, bids_root, overwrite=False)
```

satra commented:
>  but I can understand that doesn't totally specify things

@nicholst - can you unpack this a bit more? there is an entire yacc parser for cmixf-12 at the end of the document, so technically it does provide a complete representation of formatting as well.

also i know there are some additional issues in BIDS w.r.t  units that we have to tackle such as the fact that certain things (age, height, weight) represented as SI are likely to impose significant constraints on certain groups in certain countries.  (perhaps this should be a separate issue, but likely to come up in all contexts).

nicholst commented:
Hi @satra,

I should have rephrased my comment as "I interpret @sappelhoff's message to mean that MIXF doesn't totally specify things" :)

@sappelhoff: Maybe you can clarify what's missing from MIXF?

sappelhoff commented:
> Maybe you can clarify what's missing from MIXF?

I don't think that there is something missing from MIXF. And I *do* think that it may be a nice way forward to require MIXF formatting of SI units.

Yet, I see several issues that need to be discussed with more people than the usual suspects, because we have a high chance of breaking existing tools and datasets:

- Previously users have formatted SI units however they wanted, and nobody (including the validator) complained
- Now, the validator checks BIDS datasets according to rules that are not clearly specified in our specification (see my [summary post](https://github.com/bids-standard/bids-specification/pull/411#issuecomment-617036529))
- Thus, users may now start to adopt the "validator" rules, which are not clearly defined BIDS specification rules (because we are unclear in the spec)
- Hence we should clarify BIDS specification rules, and make the validator enforce these rules ... to avoid a second "breakage"

---

I agree with @nicholst that it would be nice to simply refer to another standard: "Please use units according to this" ... and even better if there is an existing validator for that standard that we can simply tie into our `bids-validator`


---

@satra yes I think that it'd be helpful to split off the discussion about age, and I am interested to hear more about why you think weight and height could be problematic to be represented in SI.

jasmainak commented on src/99-appendices/05-units.md:
```diff
@@ -54,15 +83,15 @@ Following the [International System of Units](https://en.wikipedia.org/wiki/Inte
+| [deci](https://www.wikiwand.com/en/Deci-)   | d               | 10<sup>-1</sup>  |
+| [centi](https://www.wikiwand.com/en/Centi-) | c               | 10<sup>-2</sup>  |
+| [milli](https://www.wikiwand.com/en/Milli-) | m               | 10<sup>-3</sup>  |
+| [micro](https://www.wikiwand.com/en/Micro-) | u or µ (U+00B5) | 10<sup>-6</sup>  |
```
why don't we allow both micro and mu symbols? They are indistinguishable to the naked eye and it's mostly [a historical artifact](https://en.wikipedia.org/wiki/Micro-#Symbol_encoding_in_character_sets). I think we should make it easy  for people to adapt BIDS, not make them spend hours understanding the difference between these two symbols.

sappelhoff replied:
> I think we should make it easy for people to adapt BIDS, not make them spend hours understanding the difference between these two symbols.

I agree. I guess we are balancing between narrowing down possibilities of formatting and being overly pedantic. In this case it's currently perhaps being overly pedantic 🙂 

@effigies brought this up in an online meeting not too long ago ... soo I'll update this soon, thanks @jasmainak

effigies replied:
The reason that we didn't allow both was that it shifts more of the responsibility onto apps to parse the units. If we can say they're always MIXF, then all you have to know if MIXF. We had some suggestions for micro-, degree and Ohm (IIRC) that made this difficult by backwards compatibility, so we said MIXF + those three code points. We can just as easily say MIXF + mu + micro + degree + Ohm, but there will be a complexity cost to any software that cares about units.

Right now we're seeing a case where the validator requiring micro instead of mu is causing user confusion. We can relieve this momentarily by relaxing the constraint, but will we not simply be distributing the confusion to every tool's bug tracker?

jasmainak replied:
humm ... I just went back and checked how we are dealing with this in MNE. We coerce the mu and micro into micro -- so it might be a tool-to-tool issue. But I agree that keeping one would be simpler for tool developers in the long run.

emdupre replied:
I'm with @jasmainak on this one -- the difference between these two characters is imperceptible, so I really don't like pushing this onto users.

I do understand your concern, though, @effigies. If we want to enforce this with the validator, then I think this case of mu vs micro needs a very specific message alerting users to exactly this difference. To my mind, that seems the only reasonable (humane ?) way to handle it.

What I don't know, though, is how easy this would be to accomplish with the current validator structure. WDYT ?

sappelhoff replied:
I think it should be easily possible in the current validator structure:
- implement a new issue code Z to signal "you are using µ as a character, please be aware that ... do not use U+03BC ..." 

and then for validation:

- when parsing a `unit` column in a TSV file X
- upon encountering unicode `U+03BC` in a line Y of file X
- push an issue to issue list for file X at line Y: issue code Z

what's more of a challenge (and one of the reasons why this PR is stagnating) is to validate MIXF units as a whole :confused:

sappelhoff replied:
re: validator -> @satra now has implemented a nice Python validator: https://github.com/sensein/cmixf

the JS version of a cmixf validator may follow after merging this PR, but it shouldn't be a hard blocker.

jasmainak replied:
we should raise an issue and track these somewhere nonetheless

sappelhoff replied:
agreed, I am now tracking our discussion here: https://github.com/bids-standard/bids-validator/issues/969

thanks for the many points.

nellh replied:
Why not rely on Unicode normalization to solve the visually similar character issue? For example, the compatibility decomposition will map U+00B5 to U+03BC. Python and JavaScript both have support for these mappings in the standard libraries. bids-validator could just normalize all Unicode text before validation. The specification could also recommend which normalization is preferred for other software aiming to be compatible.

sappelhoff replied:
@nellh thanks for your input and also the references (above) -> could you make a concrete proposal what needs to be changeds in this PR (using GitHub's code suggestion feature)?

gpiantoni replied:
I'd like to point out that according to the unicode standard ([PDF](https://www.unicode.org/reports/tr25/), p.11) the sign `U+03BC` is the preferred symbol. So it'd be surprising to some users to get a warning when using the preferred `U+03BC` instead of `U+00B5` (which according to unicode, should be recognized as well)

sappelhoff replied:
Thanks @gpiantoni. I take your post as voting for option 3 in my summary (https://github.com/bids-standard/bids-specification/pull/411#issuecomment-633992688), right? ... what are your thoughts about option 1?

gpiantoni replied:
Thanks for asking my opinion. As end user, I'd prefer option 3. But I understand the complexity of implementing it, so option 1 is probably easier to implement (sorry, I don't know more than this to comment further). Option 2 however would put the BIDS guidelines in disagreement with the unicode standard.

effigies commented:
@emdupre Yes, I agree that it's a pain for users and that a good error message is urgent if we do place these constraints. But a validator is pretty ideal for complaining about subtle problems and providing a suggestion, which is why I'm leaning that way rather than simply relaxing the constraint.

I would suggest that we detect any Unicode units and suggest MIXF alternatives, make them warnings for permitted code points and errors for not.

And to be clear, I'm willing to be overruled on whether mu is permitted.

rwblair commented:
I've been unable to get the yacc grammar specified in the standard to compile, with the start symbol set as quantity_value I get the following errors:
```
mixf.y: warning: 24 nonterminals useless in grammar [-Wother]
mixf.y: warning: 150 rules useless in grammar [-Wother]
mixf.y:1.8-21: fatal error: start symbol quantity_value does not derive any sentence
 %start quantity_value
```
using byacc I get the above in the out put log as well as the following:
```
byacc: 2 rules never reduced
byacc: 8 shift/reduce conflicts, 649 reduce/reduce conflicts.
```
We might be able to get an MIXF parser from the grammar in node with  https://github.com/zaach/jison. JS it produces has `require('fs')` so running in the browser version of the validator would take modification if possible at all.

Is there any other MIXF tooling out there? Implementing ourselves without an automatic parser generator might take some time.

MaxvandenBoom commented:
Hi everyone,

I am also running into the problem of microVolts units being written as μV ('GREEK SMALL LETTER MU' in Unicode) or µV ('MICRO SIGN'). The bids-validator error and the bids specification (referred to in the error) are a bit ambiguous on this issue: https://bids-specification.readthedocs.io/en/stable/99-appendices/05-units.html

I dug a bit into your discussion, and I ended up with two questions:
1. Is there a final decision yet on which "character sign" to use?
2. The BIDS-specification say JSON files can be encoded in UTF-8 or ASCII. I can't find (but I might have missed) such a statement for .tsv files. Our .tsv files are in ASCII, while the units.js is in UTF-8. I think the difference in encoding might makes the node bids-validator fail for our ASCII .tsv files, despite using the correct character (I think there is only one for mu/micro in ASCII?). Doesn't that mean that, de facto, .tsv files are required to be in UTF-8? And is that the way forward?

Thanks,
Max

sappelhoff commented:
Mmh where did you find that JSON files can be encoded in ASCII?

Let me quote from the spec:


> TSV files MUST be in UTF-8 encoding.

see: https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#tabular-files

and

> JSON files MUST be in UTF-8 encoding.

https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#keyvalue-files-dictionaries

sappelhoff commented:
> Is there a final decision yet on which "character sign" to use?

For your case it'll be safe to use the MICRO SIGN (µ), because however the discussion may evolve, that "character sign" will be supported for backward compatibility reasons.

MaxvandenBoom commented:
@sappelhoff You are perfectly right, I only found the statement about the JSON files, but from a different source. I will correct to UTF-8 and use the MICRO SIGN. Thank you for the clarification!

MaxvandenBoom commented:
@sappelhoff The bids-validator gave me an error when JSON files are not in UTF-8 (`"JSON files must be valid utf-8."`). But it did not give me a similar error for `.tsv` files, this is one of the reasons I thought ASCII was ok.

Could it be that such an UTF-8 check is not implemented for .tsv files?

sappelhoff commented:
> Could it be that such an UTF-8 check is not implemented for .tsv files?

very possible :thinking: can you open an issue for this? Perhaps we can implement it if it's not there yet.

MaxvandenBoom commented:
> > Could it be that such an UTF-8 check is not implemented for .tsv files?
> 
> very possible  can you open an issue for this? Perhaps we can implement it if it's not there yet.

Definitely! It is here now https://github.com/bids-standard/bids-validator/issues/962

sappelhoff commented on src/02-common-principles.md:
```diff
@@ -377,7 +377,9 @@ data is expressed in SI units or SI derived units, the units MAY be specified in
-prefixes.
+prefixes. Note also that for the *formatting* of SI units, the
+[CMIXF-12](https://people.csail.mit.edu/jaffer/MIXF/CMIXF-12) convention for
+encoding units is RECOMMENDED.
```
```suggestion
encoding units is REQUIRED.
```

jasmainak replied:
can we add an example here? What would be CMIXF-12 compatible and what would not be? Just to make sure that users don't get lost in trying to understand another standard ...

sappelhoff replied:
Good point, ... there are some examples in the appendix, but I'll also add some here. :+1:

sappelhoff commented on src/99-appendices/05-units.md:
```diff
@@ -1,39 +1,68 @@
-| sievert        | Sv          | equivalent dose (of ionizing radiation)    |
-| katal          | kat         | catalytic activity                         |
+The [CMIXF-12](https://people.csail.mit.edu/jaffer/MIXF/CMIXF-12) convention
+for encoding units is RECOMMENDED to achieve maximum portability and limited
```
```suggestion
for encoding units is REQUIRED to achieve maximum portability and limited
```

sappelhoff commented on src/99-appendices/05-units.md:
```diff
@@ -1,39 +1,68 @@
-| katal          | kat         | catalytic activity                         |
+The [CMIXF-12](https://people.csail.mit.edu/jaffer/MIXF/CMIXF-12) convention
+for encoding units is RECOMMENDED to achieve maximum portability and limited
+variability of representation.
```
```suggestion
variability of representation.
In case that a CMIXF representation of SI units is not possible, you will have
to declare your units as custom units and define them in accompanying JSON
files, as described in the [units section](../02-common-principles.md#units).
```

satra commented:
just an fyi, i implemented a python parser here: https://github.com/sensein/cmixf (still has a bunch of corner cases x-failed), and one could probably use a different lexer/parser library. to be used in practice those corner cases need to be resolved.

sappelhoff commented:
> just an fyi, i implemented a python parser here: https://github.com/sensein/cmixf

that's really cool :+1: unfortunately I don't think that we can integrate it into the bids-validator :thinking: at least not for the web-version. But perhaps somebody can take it from there and translate it to JS. Not sure how difficult that would be. cc @rwblair 

But I am now tending towards not seeing the validation as a blocker for merging the PR. I think the validator can follow the spec (preferably quickly) in this.

satra commented:
>But I am now tending towards not seeing the validation as a blocker for merging the PR. I think the validator can follow the spec (preferably quickly) in this.

+1 to that. the intent with the library was two-fold. not too hard to implement a parser and being pure python there are many ways to convert to js straight up that could be included in the bids-validator, but also used by other python libraries. (btw, thanks for the feedback - will see if i can fix those issues)

sappelhoff commented:
Okay, I feel like this is ready for a final review now. @bids-standard/bep_leads @bids-maintenance @bids-standard/raw

effigies commented on src/02-common-principles.md:
```diff
@@ -413,16 +413,21 @@ label, but must be included in file names (similarly to other key names).
+convention for encoding units is REQUIRED.
+More information about CMIXF can be found in [Appendix V](99-appendices/05-units.md)
+but to provide some examples, CMIXF formatting of "micro volts" would be `uV`,
+and of "kilobecquerel per ml" would be `kBq/ml`.
```
```suggestion
and of "kilobecquerel per milliliter" would be `kBq/mL`.
```

effigies commented on src/99-appendices/05-units.md:
```diff
@@ -1,39 +1,76 @@
+| Unit name      | Unit symbol       | Quantity name                              |
+| -------------- | ----------------- | ------------------------------------------ |
+| metre          | m                 | length                                     |
+| kilogram       | kg                | mass                                       |
```
We're missing liters.

```suggestion
| kilogram       | kg                | mass                                       |
| litre (liter)  | L                 | volume                                     |
```

effigies commented on src/99-appendices/05-units.md:
```diff
@@ -1,39 +1,76 @@
+
+It is REQUIRED that units be CMIXF-12 compliant or among these three Unicode
+characters.
+Please take care to not confuse the accepted unicode `U+00B5` (µ) with the
+unicode NOT accepted `U+03BC` (μ):
```
Minor grammar issues, and added links to a useful reference that can give additional context.

```suggestion
Please take care not to confuse the accepted Unicode character
[`U+00B5`](https://codepoints.net/U+00B5) (µ) with the disallowed Unicode
character [`U+03BC`](https://codepoints.net/U+03BC) (μ):
```

effigies commented on src/99-appendices/05-units.md:
```diff
@@ -1,39 +1,76 @@
+The [CMIXF-12](https://people.csail.mit.edu/jaffer/MIXF/CMIXF-12) convention
+for encoding units is REQUIRED to achieve maximum portability and limited
+variability of representation.
+In case that a CMIXF representation of SI units is not possible, you will have
+to declare your units as custom units and define them in accompanying JSON
+files, as described in the [units section](../02-common-principles.md#units).
```
Eliminate "you" phrasing:

```suggestion
If a CMIXF-12 representation of a unit is not possible, the unit MUST be declared
as custom units and defined in an accompanying JSON file, as described in the
[units section](../02-common-principles.md#units).
```

jasmainak commented on src/02-common-principles.md:
```diff
@@ -413,16 +413,21 @@ label, but must be included in file names (similarly to other key names).
+convention for encoding units is REQUIRED.
+More information about CMIXF can be found in [Appendix V](99-appendices/05-units.md)
+but to provide some examples, CMIXF formatting of "micro volts" would be `uV`,
+and of "kilobecquerel per milliliter" would be `kBq/mL`.
```
I think we should also give an example of what would be an invalid CMIXF formatting.

sappelhoff replied:
There is:

 1.  `uV` or `µV` are permissible, but NOT: `microV`, `µvolt`, `1e-6V`, etc.

a bit further below, I think that suffices, but feel free to make me do it if you feel strongly about it

jasmainak replied:
yeah I noticed once I read the whole thing. I think CMIXF comes out of the blue a bit. If there was an example to motivate why to use CMIXF it may help. But it's up to you.

jasmainak commented on src/02-common-principles.md:
```diff
@@ -435,17 +440,24 @@ For additional rules, see below:
-    remember to shift dates within one subject by the same number of days to
-    maintain the interval information. Example: `1867-06-15T13:45:30`
+    reasons.
+    To distinguish real dates from shifted dates always use year 1925
```
```suggestion
    To distinguish real dates from shifted dates, always use year 1925
```

jasmainak commented on src/99-appendices/05-units.md:
```diff
@@ -1,39 +1,78 @@
+[CMIXF-12](https://people.csail.mit.edu/jaffer/MIXF/CMIXF-12) convention
+applies.
+You can use the [cmixf Python package](https://github.com/sensein/cmixf) to
+check whether your formatting is compliant.
```
is there a corresponding issue on the BIDS-validator? Ideally, the validator should tell you instead of having to use another tool. Also works for people who are not familiar with Python.

sappelhoff replied:
Not yet, but @rwblair is aware of it and I'll open an issue as soon as this PR is in to continue tracking this challenge.

sappelhoff replied:
now there is an issue: https://github.com/bids-standard/bids-validator/issues/969

MaxvandenBoom commented:
@sappelhoff The text now reads `u or µ (U+00B5)`. I am not entirely certain, but I think that `00B5` is the UTF-16+ representation. It think in UTF-8 the representation in bytes is `0xC2 0xB5 (c2b5)`, at least that gives me the correct UTF-8 output, which is accepted by the validator.

I know it might depend on the programming language and text writing algorithm how it handles UTF; but it might be confusing to mention only the UTF-16 representation in the documentation, while UTF-8 encoding is required.

sappelhoff commented on src/99-appendices/05-units.md:
```diff
@@ -1,39 +1,78 @@
+Earlier versions of the BIDS standard listed the following Unicode symbols, and
+these are still included for backwards compatibility:
+
+1.  `U+00B5` (µ)
+1.  `U+00B0` (°)
+1.  `U+2126` (Ω).
```
```suggestion
1.  [`U+00B5` (µ)](https://codepoints.net/U+00B5)
1.  [`U+00B0` (°)](https://codepoints.net/U+00B0)
1.  [`U+2126` (Ω)](https://codepoints.net/U+2126)
```

nellh commented on src/99-appendices/05-units.md:
```diff
@@ -1,39 +1,78 @@
+Please take care not to confuse the accepted Unicode character
+[`U+00B5`](https://codepoints.net/U+00B5) (µ) with the disallowed Unicode
+character [`U+03BC`](https://codepoints.net/U+03BC) (μ):
+They look equal in appearance but are different to the computer.
```
Since U+00B5 maps to U+03BC but not the other way around, shouldn't U+03BC be the preferred value? That way data files which are modified by software using the compatibility mapping don't need to be manually fixed after they are modified.

sappelhoff replied:
thanks @nellh that sounds reasonable - I have not heard of this mapping before -> do you have a reference where I could read up on this?

nellh replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize for JavaScript.

https://docs.python.org/3/library/unicodedata.html#unicodedata.normalize for Python.

Wikipedia also has a general summary under https://en.m.wikipedia.org/wiki/Unicode_equivalence

sappelhoff replied:
> Since U+00B5 maps to U+03BC but not the other way around

Okay I get it now. Here is a Python example:

```Python
import unicodedata

# get utf-8 encoding of MICRO sign after NFC norm
unicodedata.normalize('NFC', 'µ').encode()                             
b'\xc2\xb5'

# same but NFKC norm --> MICRO sign is now SMALL GREEK MU
unicodedata.normalize('NFKC', 'µ').encode()                           
b'\xce\xbc'

# try same for SMALL GREEK MU: NFC norm doesn't change anything
unicodedata.normalize('NFKC', 'μ').encode()                           
b'\xce\xbc'

# also NFKC norm doesn't change anything
unicodedata.normalize('NFC', 'μ').encode()                            
b'\xce\xbc'
```

sappelhoff commented:
addressing review by @nellh --> which exact unicodes should be supported, and which should not?

for context, see:

- https://github.com/bids-standard/bids-specification/pull/411#discussion_r430370932
- https://github.com/bids-standard/bids-specification/pull/411/files#r430256466

Regarding this whole discussion I want to stress that **preferably** we would completely cut out µ, °, and  Ω. Unfortunately, we have to keep them for backward compatibility reasons (but see option 1 below).

µ has a particularly confusing history, because we used to have GREEK SMALL MU in the BIDS appendix first, which was then changed to MICRO later in November 2018: https://github.com/bids-standard/bids-specification/pull/73

Looking through that PR once more also alerts me to the fact that for Ω there are also several unicodes: [OHM](https://www.compart.com/en/unicode/U+2126) and [GREEK CAPITAL OMEGA](https://www.compart.com/en/unicode/U+03A9)

To move forward with this PR, I see these three options:

1. disregard backward compatibility and stop support for the three unicodes
1. See https://github.com/bids-standard/bids-specification/pull/73 as a bugfix, and only accept the unicodes `U+00B5`, `U+00B0`, `U+2126` (as is currently advertised)
1. Realize that over the course of BIDS we have accepted all kinds of unicodes for the three symbols °, µ, and Ω ... and now start dealing with it by accepting them in any form (e.g., both MICRO and GREEK SMALL MU are valid) ... and then tell all downstream software to convert these chars according to "compatibility decomposition" (e.g., MICRO gets mapped to SMALL GREEK MU, but SMALL GREEK MU stays what it is) etc.

What are your opinions?

nicholst commented:
I would cast a vote for option 1, "disregard backward compatibility".  While I realise this is generally not a good action, I'm thinking:

* I assume that there are relatively few documents that this would turn into "not valid" BIDS documents in terms of a validator, *and*
* Even such "not valid" BIDS datasets would be completely usable for all uses I can think of, and
* This action provides easier support in the future by limited the character set used.

alexrockhill commented:
> I would cast a vote for option 1, "disregard backward compatibility". While I realise this is generally not a good action, I'm thinking:
> 
> * I assume that there are relatively few documents that this would turn into "not valid" BIDS documents in terms of a validator, _and_
> * Even such "not valid" BIDS datasets would be completely usable for all uses I can think of, and
> * This action provides easier support in the future by limiting the character set used.

Maybe it wouldn't be too hard to parse through the OpenNeuro datasets and change them solving the backward compatibility issue for publicly available datasets. I'd volunteer to look through and fix the datasets I've uploaded.

nellh commented:
For units I agree with #1. Making them easier to parse without character choice corner cases simplifies a lot for a relatively minor compatibility break. Maybe bids-validator could suggest automatically fixing the common case when it raises an error for this?

For Unicode text in general, there would be some value in specifying how it should be normalized in the BIDS spec. Agreeing on the method of normalization would help resolve some of these visually similar characters or non-canonical codepoint issues for text other than units by giving tools interpreting text a set of guidelines to follow.

sappelhoff commented:
Thanks for your input and perspectives everybody!

We just discussed this in a maintainers meeting and agreed that we will act in a backward compatible manner for this case in order to keep breakage to a minimum.

Concretely this means I have implemented option 3. from my summary:

1.  [`U+03BC` (μ)](https://codepoints.net/U+03BC) or [`U+00B5` (µ)](https://codepoints.net/U+00B5)
1.  [`U+03A9` (Ω)](https://codepoints.net/U+03A9) or [`U+2126` (Ω)](https://codepoints.net/U+2126)
1.  [`U+00B0` (°)](https://codepoints.net/U+00B0)

are now permitted Unicodes. I also link to the Unicode technical report section that explains the duplicated characters for micro and ohm.

Lastly, I removed micro and ohm from the table below, because I see no need to advertise this option: We keep it only for compatibility purposes -> new datasets should adopt CMIXF convention. (also, the table would start to look ugly if we would include three unicodes there while in all other columns it's only 1 to 3 letters)

From my perspective, this issue is now solved and the review can continue. Feel free to raise this specific point again if I did not explain clearly enough.

rwblair commented:
@satra Nice work on the parser. I took a swing at using the lex file from your package with jison to generate a javascript parser: https://github.com/poldracklab/cmixf-js

At some point it would be good to have a shared list of good and bad examples of cmixf strings to validate against to try and catch any differences between the python and js versions.

sappelhoff commented:
Roadmap: I plan to merge this PR before the next release on the 6th of May 2020.

Please raise remaining concerns before that so we can discuss and adjust.

Unfortunately several approving review were dismissed by the changes in this PR. Reviewers, could you please check the updates and re-approve?

effigies commented:
Given that there are two approvals, I want to put the brakes on, because I think we're inconsistent about RECOMMENDED/REQUIRED usage.

I'll try to get my comments in line today.

effigies commented:
Just to be clear, my interpretation of the consensus is that there are two cases, and we just need to communicate them clearly and consistently:

1) A unit is defined as an SI or SI-derived unit, encoded in CMIXF-12. (A discouraged but backward-compatible Unicode exception exists.)
2) A custom unit is defined in JSON, and a validator need only check it exists.

Assuming we are indeed agreed, then I can submit suggestions for wording later. But I see two additional problems.

The smaller is that, of the four examples of units we have, one is SI+MIXF, one is SI+Unicode, and two are custom. The bigger problem is that we have no clear description or example of specifying a custom unit.

Cases:

1) In the [Tabular files](https://2602-150465237-gh.circle-artifacts.com/0/site/02-common-principles.html#tabular-files) section, we say that SI units in CMIXF are "REQUIRED if possible" and the (long-standing) example is "kilograms per squared meters"

https://github.com/bids-standard/bids-specification/blob/6a88b04e2212c08469ef5819a8d1eba5558feea5/src/02-common-principles.md#L348-L352

While we could change this to `"kg/m^2"`, it leaves open the question: What happens with the "not possible" case? This is the JSON sidecar that should presumably be defining custom units. What would that look like?

2) [MEG channels](https://bids-specification.readthedocs.io/en/latest/04-modality-specific-files/02-magnetoencephalography.html#channels-description-_channelstsv). Example units are `V` and `T`. Valid.

3) [EEG channels](https://bids-specification.readthedocs.io/en/latest/04-modality-specific-files/03-electroencephalography.html#channels-description-_channelstsv). Example shows `microV`. Obviously, we should switch to `uV`, but again, what if a custom unit were necessary? Where is the unit to be defined?

4) [iEEG channels](https://bids-specification.readthedocs.io/en/latest/04-modality-specific-files/04-intracranial-electroencephalography.html#channels-description-_channelstsv). Example shows `μV`. Valid, but we should probably change.

---

Should we just disallow custom units? Or should we come up with a structure for defining them?

robertoostenveld commented:
thanks for finding and pointing out the inconsistencies in the current spec. On the EEG page, I also noticed another `μV`. 

I agree with your interpretation.

- I recommend updating all examples to reflect the latest insights. 
- I think we should allow custom units. E.g. when studying neonates, it would make sense to indicate age in weeks, which is not a SI unit. Other reasonably well-defined but non-SI units could be parsec (unlikely to be used in BIDS), or frames (related to a stimulus on screen). Others are feet, pounds 
- I don't think we should bother with specifying how custom units should be defined (also considering the Pareto principle).

satra commented:
@effigies - the discussion of custom units was considered out of scope for this PR in a previous PR/issue/comment.

effigies commented:
I'm not sure how we can declare any formatting required (in a validatable sense) if there is no way to invoke the custom units clause.

I think at best we can say these are recommended, and the validator can warn.

satra commented:
> I'm not sure how we can declare any formatting required (in a validatable sense) if there is no way to invoke the custom units clause.

yes, that i suspect is the result of a change with the wording from recommended to required. so the classic outliers are:

- age/period (non SI, could potentially use https://en.wikipedia.org/wiki/ISO_8601#Durations + reference)
- weight (would we not allow lbs, oz?)
- length/height/distance (miles, yards, ft, inches vs metric)

the simplest changecould be an option like this:
1. explicit key (one of Units or CustomUnits is required)
```
"CustomUnits": string
```
or
2. still use the units key but allow a set of custom units for  the classic categories

sappelhoff commented:
> I'm not sure how we can declare any formatting required (in a validatable sense)

When I made this change I was thinking about the electrophys `channels.tsv` files that have a `units` column. The validator logic would have been:

- check `units` column ... if SI, all is fine
- if not SI ... check if there is accompanying channels.JSON with `units` field
- if there is, assume custom units are defined, all is fine
- else, raise error (because no SI, and no JSON with units field are found)

I realize now that this was probably too electrophys centric and there will be cases when units are defined only in a JSON (without any TSV even necessary)

> the simplest change could be an option like this: explicit key (one of `Units` or `CustomUnits` is required)

I like this option because it would allow us to truly REQUIRE (and validate) SI units in the "units" field, while still preserving a lot of flexibility for other usecases

We can then have a short section with examples for custom units.

> I think at best we can say these are recommended, and the validator can warn.

This is the alternative option (instead of introducing `CustomUnits`). This would make the whole attempt to better standardize units within BIDS a bit less strong ... but one might argue that the steps we take here are already enough with the RECOMMENDED keyword, as opposed to using the REQUIRED keyword

Which of the options do you prefer @effigies @robertoostenveld @nicholst ?

sappelhoff commented:
PS: Just found another custom unit: "pixels" as an accepted unit for the `iEEGCoordinateUnits` field in the [coords section](https://bids-specification.readthedocs.io/en/latest/04-modality-specific-files/04-intracranial-electroencephalography.html#coordinate-system-json-_coordsystemjson)

This would rather be an argument for the "do not require, but recommend and warn" route. --> and then have a list of units where we don't warn (pixels being one entry in that list)

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -537,8 +537,11 @@ sub-<label>/[ses-<label>/]
+coregistration.
+In this case the sidecar JSON file needs to include the units of the fieldmap.
+The possible options are: Hertz (`Hz`), Radians per second (`rad/s`), or Tesla
+(`T`).
```
the previous specification text may be interpreted such that `Tesla` is a permitted value. That would not be true anymore with our CMIXF changes.

sappelhoff commented on src/04-modality-specific-files/03-electroencephalography.md:
```diff
@@ -376,7 +376,7 @@ Fields relating to the EEG electrode positions:
 | --------------------------------------------------------------------------------------------------------------------------------------------------| --------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
 | EEGCoordinateSystem            | REQUIRED. Refers to the coordinate system in which the EEG electrode positions are to be interpreted (see [Appendix VIII](../99-appendices/08-coordinate-systems.md)). |
-| EEGCoordinateUnits             | REQUIRED. Units in which the coordinates that are listed in the field `EEGCoordinateSystem` are represented (e.g., "mm", "cm").                                        |
+| EEGCoordinateUnits             | REQUIRED. Units in which the coordinates that are listed in the field `EEGCoordinateSystem` are represented. MUST be `m`, `cm`, or `mm`.                               |
```
I see this change as a clarification / issue fix, because the EEG spec were to follow the MEG spec in this regard, and MEG clearly states: 

> MUST be `m`, `cm`, or `mm`

[see coords section](https://bids-specification.readthedocs.io/en/latest/04-modality-specific-files/02-magnetoencephalography.html#coordinate-system-json-_coordsystemjson)

This change is also applied to `FiducialsCoordinateUnits` and `AnatomicalLandmarkCoordinateUnits`

yet ... in the MRI [`AnatomicalLandmarkCoordinateUnits` section](https://bids-specification.readthedocs.io/en/latest/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#anatomical-landmarks), voxels are given as the unit to be used --> So I wonder whether we should add `voxels` as permitted in EEG and MEG as well

allowing `voxels` on top of m, cm, and mm would also correspond nicely to the allowed `pixels` unit in iEEG for these fields.

sappelhoff commented on src/04-modality-specific-files/04-intracranial-electroencephalography.md:
```diff
@@ -339,13 +339,13 @@ listed below.
+| Column name  | Definition                                                                                                                                 |
+| ---------------------- | -------------------------------------------------------------------------------------------------------------------------------- |
+| name         | REQUIRED. Name of the electrode contact point.                                                                                             |
+| x            | REQUIRED. X position. The positions of the center of each electrode in xyz space. Units are specified in `*_space-<label>_electrode.json`. |
```
```suggestion
| x            | REQUIRED. X position. The positions of the center of each electrode in xyz space. Units are specified in `*_space-<label>coordsystem.json`. |
```

I think this may be an error in the iEEG spec. AFAIK this info should be specified in COORDSYSTEM.json, ...

robertoostenveld replied:
agreed. Note that you have a _ missing in the corrected `*_space-<label>_coordsystem.json`

nicholst commented on src/02-common-principles.md:
```diff
@@ -413,16 +413,21 @@ label, but must be included in file names (similarly to other key names).
 prefixes.
+Note also that for the *formatting* of SI units, the [CMIXF-12](https://people.csail.mit.edu/jaffer/MIXF/CMIXF-12)
+convention for encoding units is REQUIRED.
+More information about CMIXF can be found in [Appendix V](99-appendices/05-units.md)
+but to provide some examples, CMIXF formatting of "micro volts" would be `uV`,
+and of "kilobecquerel per milliliter" would be `kBq/mL`.
```
Well, if we're not in a rush to approve, can I suggest smoother wording here? "... but to provide..." is weird. Better would be to sketch out what CMIXF is trying to do, then provide an example; also, "kBq/mL" isn't so informative b/c that's a routine abbreviation -- better to focus on the exceptions that might be counter intuitive.

What about this text for this passage, starting after "convention for encoding units is REQUIRED."

> CMIXF provides a consistent system for all units and prefix symbols with only basic characters, avoiding symbols that can cause text encoding problems; for example the CMIXF formatting for "micro volts" is `uV`, "degrees Celsius" is `oC` and "Ohm" is `Ohm`. See [Appendix V](99-appendices/05-units.md) for more information.

Happy to make a PR if people like this.

sappelhoff replied:
> if we're not in a rush to approve

definitely not in a rush, I am just trying to keep the momentum up so that this PR does not go stale for another year. If we see problems or ways to make it better, we should (and will) take the time to make it as good as we can given all constraints.

> Happy to make a PR if people like this.

PRs are welcome to: https://github.com/sappelhoff/bids-specification/tree/asciiunits

an alternative for PRs is to use the [GitHub suggestion function](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/reviewing-proposed-changes-in-a-pull-request) --> multiline suggestions are now also possible [gh-blog-post on this](https://github.blog/changelog/2020-02-26-multi-line-code-suggestions-beta/)

nicholst replied:
Cool!  [Done](https://github.com/bids-standard/bids-specification/pull/411#discussion_r432484588).

sappelhoff replied:
Thanks, this is a lot better.

nicholst commented on src/02-common-principles.md:
```diff
@@ -413,16 +413,21 @@ label, but must be included in file names (similarly to other key names).
+non-standard prefixed units MUST be specified in the JSON file.
 See [Appendix V](99-appendices/05-units.md) for a list of standard units and
 prefixes.
+Note also that for the *formatting* of SI units, the [CMIXF-12](https://people.csail.mit.edu/jaffer/MIXF/CMIXF-12)
+convention for encoding units is REQUIRED.
+More information about CMIXF can be found in [Appendix V](99-appendices/05-units.md)
+but to provide some examples, CMIXF formatting of "micro volts" would be `uV`,
+and of "kilobecquerel per milliliter" would be `kBq/mL`.
```
```suggestion
CMIXF provides a consistent system for all units and prefix symbols with only basic characters, avoiding symbols that can cause text encoding problems; for example the CMIXF formatting for "micro volts" is uV, "degrees Celsius" is oC and "Ohm" is Ohm. See Appendix V for more information.
```

nicholst replied:
```suggestion
CMIXF provides a consistent system for all units and prefix symbols with only basic
characters, avoiding symbols that can cause text encoding problems; for example the
CMIXF formatting for "micro volts" is uV, "degrees Celsius" is oC and "Ohm" is Ohm. 
See Appendix V for more information.
```

sappelhoff commented:
I went through the PR once more and changed the REQUIRED keywords back to RECOMMENDED.

Now the PR achieves the following:

1. RECOMMEND CMIXF-12 units
1. provide examples of CMIXF-12 and say why we want it
1. align present spec examples to our new recommendation
1. more consistently link to the units section instead of specifying how to define units in different places
1. Clarify that some unicodes (5, to be exact) are backward compatible

From what I see, we are not breaking any previous versions of BIDS. Only making some recommendations where previously we were unclear.

This PR is now ready to be merged from my side. Please review and approve ... or tell me that I overlooked something. And thanks for your patience :+1:

effigies commented on src/02-common-principles.md:
```diff
@@ -423,13 +423,13 @@ dictionary matches a column name in the TSV file, then that field MUST contain a
-| Levels      | For categorical variables: a dictionary of possible values (keys) and their descriptions (values).                                                          |
-| Units       | Measurement units. `[<prefix symbol>] <unit symbol>` format following the SI standard is RECOMMENDED (see [units section](./02-common-principles.md#units). |
-| TermURL     | URL pointing to a formal definition of this type of data in an ontology available on the web.                                                               |
+| Field name  | Definition                                                                                                             |
+| :---------- | :--------------------------------------------------------------------------------------------------------------------- |
+| LongName    | Long (unabbreviated) name of the column.                                                                               |
+| Description | Description of the column.                                                                                             |
+| Levels      | For categorical variables: a dictionary of possible values (keys) and their descriptions (values).                     |
+| Units       | Measurement units. SI units in CMIXF formatting are RECOMMENDED (see [units section](./02-common-principles.md#units). |
+| TermURL     | URL pointing to a formal definition of this type of data in an ontology available on the web.                          |
```
```suggestion
| Field name  | Definition                                                                                                      |
| :---------- | :-------------------------------------------------------------------------------------------------------------- |
| LongName    | Long (unabbreviated) name of the column.                                                                        |
| Description | Description of the column.                                                                                      |
| Levels      | For categorical variables: a dictionary of possible values (keys) and their descriptions (values).              |
| Units       | Measurement units. SI units in CMIXF formatting are RECOMMENDED (see [Units](./02-common-principles.md#units)). |
| TermURL     | URL pointing to a formal definition of this type of data in an ontology available on the web.                   |
```

effigies commented on src/03-modality-agnostic-files.md:
```diff
@@ -336,12 +336,8 @@ imaging acquisition sequence (each run `.nii[.gz]` file) within one session.
-format, time zone is always assumed as local time).
-No specific precision is required for fractional seconds, but the precision
-SHOULD be consistent across the dataset
+Datetime should be expressed as described in the
+[Units section](./02-common-principles.md#units).
```
```suggestion
Datetime should be expressed as described in [Units](./02-common-principles.md#units).
```

effigies commented on src/04-modality-specific-files/02-magnetoencephalography.md:
```diff
@@ -199,11 +199,11 @@ The columns of the Channels description table stored in `*_channels.tsv` are:
-| name        | REQUIRED. Channel name (e.g., MRT012, MEG023)                                                                                                                                                                                                                                                                                                                                                   ...
-| type        | REQUIRED. Type of channel; MUST use the channel types listed below.                                                                                                                                                                                                                                                                                                                             ...
-| units       | REQUIRED. Physical unit of the value represented in this channel, e.g., V for Volt, specified according to the [SI unit symbol](https://en.wikipedia.org/wiki/International_System_of_Units#Base_units) and possibly prefix symbol (e.g., mV, μV), or as a [derived SI unit](https://en.wikipedia.org/wiki/SI_derived_unit) (e.g., fT/cm). For guidelines for Units and Prefixes see [Appendix  ...
+| Column name | Definition                                                                                                                                                                              |
+| ----------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| name        | REQUIRED. Channel name (e.g., MRT012, MEG023)                                                                                                                                           |
+| type        | REQUIRED. Type of channel; MUST use the channel types listed below.                                                                                                                     |
+| units       | REQUIRED. Physical unit of the value represented in this channel, e.g., `V` for Volt, or `fT/cm` for femto Tesla per centimeter (see [units section](../02-common-principles.md#units). |
```
```suggestion
| Column name | Definition                                                                                                                                                                       |
| ----------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| name        | REQUIRED. Channel name (e.g., MRT012, MEG023)                                                                                                                                    |
| type        | REQUIRED. Type of channel; MUST use the channel types listed below.                                                                                                              |
| units       | REQUIRED. Physical unit of the value represented in this channel, e.g., `V` for Volt, or `fT/cm` for femto Tesla per centimeter (see [Units](../02-common-principles.md#units)). |
```

effigies commented on src/04-modality-specific-files/03-electroencephalography.md:
```diff
@@ -214,11 +214,11 @@ The columns of the Channels description table stored in `*_channels.tsv` are:
-| name        | REQUIRED. Channel name (e.g., FC1, Cz)                                                                                                                                                                                                                                                                                                                                                          ...
-| type        | REQUIRED. Type of channel; MUST use the channel types listed below.                                                                                                                                                                                                                                                                                                                             ...
-| units       | REQUIRED. Physical unit of the value represented in this channel, e.g., V for Volt, specified according to the [SI unit symbol](https://en.wikipedia.org/wiki/International_System_of_Units#Base_units) and possibly prefix symbol (e.g., mV, μV), or as a [derived SI unit](https://en.wikipedia.org/wiki/SI_derived_unit) (e.g., fT/cm). For guidelines for Units and Prefixes see [Appendix  ...
+| Column name | Definition                                                                                                                                                                              |
+| ----------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| name        | REQUIRED. Channel name (e.g., FC1, Cz)                                                                                                                                                  |
+| type        | REQUIRED. Type of channel; MUST use the channel types listed below.                                                                                                                     |
+| units       | REQUIRED. Physical unit of the value represented in this channel, e.g., `V` for Volt, or `fT/cm` for femto Tesla per centimeter (see [units section](../02-common-principles.md#units). |
```
```suggestion
| Column name | Definition                                                                                                                                                                       |
| ----------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| name        | REQUIRED. Channel name (e.g., FC1, Cz)                                                                                                                                           |
| type        | REQUIRED. Type of channel; MUST use the channel types listed below.                                                                                                              |
| units       | REQUIRED. Physical unit of the value represented in this channel, e.g., `V` for Volt, or `fT/cm` for femto Tesla per centimeter (see [Units](../02-common-principles.md#units)). |
```

effigies commented on src/04-modality-specific-files/04-intracranial-electroencephalography.md:
```diff
@@ -235,13 +235,13 @@ The columns of the Channels description table stored in \*\_channels.tsv are:
+| ------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| name         | REQUIRED. Label of the channel. The label must correspond to \_electrodes.tsv name and all ieeg type channels are required to have a position. The reference channel name MAY be provided in the reference column.    |
+| type         | REQUIRED. Type of channel, see below for adequate keywords in this field.                                                                                                                                             |
+| units        | REQUIRED. Physical unit of the value represented in this channel, e.g., `V` for Volt, or `fT/cm` for femto Tesla per centimeter (see [units section](../02-common-principles.md#units).                               |
```
```suggestion
| units        | REQUIRED. Physical unit of the value represented in this channel, e.g., `V` for Volt, or `fT/cm` for femto Tesla per centimeter (see [Units](../02-common-principles.md#units)).                                      |
```

effigies commented on src/99-appendices/05-units.md:
```diff
@@ -1,12 +1,56 @@
 
-Following the [International System of Units](https://en.wikipedia.org/wiki/International_System_of_Units)
-(SI, abbreviated from the French Système international (d'unités))
+As described in the [units section](../02-common-principles.md#units),
```
```suggestion
As described in the [Units](../02-common-principles.md#units),
```

sappelhoff commented on src/02-common-principles.md:
```diff
@@ -323,13 +323,13 @@ dictionary matches a column name in the TSV file, then that field MUST contain a
+| LongName    | Long (unabbreviated) name of the column.                                                                                          |
+| Description | Description of the column.                                                                                                        |
+| Levels      | For categorical variables: a dictionary of possible values (keys) and their descriptions (values).                                |
+| Units       | Measurement units. SI units in CMIXF formatting are REQUIRED when possible (see [units section](./02-common-principles.md#units). |
```
could it be that you performed parts of your review on an outdated commit @effigies?

effigies commented:
Sorry, yes. The "Review" button did not show that I had these in queue, I guess because they were outdated. I'll delete them.

sappelhoff commented:
I will merge this PR early next week, so everybody has some more time to complain or approve ... a bit like they say at some weddings:

> Speak now or forever hold your peace

sappelhoff commented:
Thanks a lot to everybody who helped along the way. Especially to @satra who brought this up.
